language/functional_tests/tests/testsuite.rs

```diff
@@ -2,28 +2,191 @@
 fn functional_tests(path: &Path) -> datatest_stable::Result<()> {
     let input = read_to_string(path)?;
 
-    let (config, directives, transactions) = split_input(&input)?;
+    let lines: Vec<String> = input.lines().map(|line| line.to_string()).collect();
+
+    let (config, directives, transactions) = split_input(&lines)?;
     let config = GlobalConfig::build(&config)?;
     let transactions = build_transactions(&config, &transactions)?;
 
     let log = eval(&config, &transactions)?;
-    if let Err(e) = check(&log, &directives) {
-        // TODO: allow the user to select debug/display mode
-        println!("{}", log);
-        return Err(e.into());
+
+    let res = match_output(&log, &directives);
+
+    let errs = match res.status {
+        MatchStatus::Success => return Ok(()),
+        MatchStatus::Failure(errs) => errs,
+    };
+
+    let mut stdout = StandardStream::stdout(ColorChoice::Auto);
+
+    macro_rules! print_directive {
+        ($idx: expr) => {{
+            let d = &directives[$idx];
+            write!(stdout, "{} | {}", d.line + 1, &lines[d.line][..d.start])?;
+            stdout.set_color(ColorSpec::new().set_underline(true))?;
+            write!(stdout, "{}", &lines[d.line][d.start..d.end])?;
+            stdout.reset()?;
+            write!(stdout, "{}", &lines[d.line][d.end..])
+        }};
     }
-    Ok(())
+
+    macro_rules! print_match {
+        ($indent: expr, $is_positive: expr, $m: expr) => {{
+            let m: &Match = $m;
+            let indent: &str = $indent;
+            let prefix = format!("[{}] ", m.entry_id);
+            let (before, matched, after) =
+                at_most_n_before_and_m_after(&res.text[m.entry_id], 30, m.start, m.end, 50);
+            write!(stdout, "{}", indent)?;
+            write!(stdout, "{}{}", prefix, before)?;
+            stdout.set_color(ColorSpec::new().set_underline(true).set_fg(Some(
+                if $is_positive {
+                    Color::Green
+                } else {
+                    Color::Red
+                },
+            )))?;
+            write!(stdout, "{}", matched)?;
+            stdout.reset()?;
+            writeln!(stdout, "{}", after)?;
+
+            let offset = prefix.chars().count() + before.chars().count();
+            write!(stdout, "{}", indent)?;
+            write!(
+                stdout,
+                "{}",
+                iter::repeat(' ').take(offset).collect::<String>()
+            )?;
+            print_directive!(m.pat_id)?;
+            writeln!(stdout)
+        }};
+    }
+
+    writeln!(stdout)?;
+
+    stdout.set_color(ColorSpec::new().set_bold(true).set_fg(Some(Color::Yellow)))?;
+    write!(stdout, "info: ")?;
+    stdout.set_color(ColorSpec::new().set_bold(true))?;
+    writeln!(stdout, "Evaluation Outputs")?;
+    stdout.reset()?;
+    let pretty = env::var("PRETTY")
+        .unwrap_or_else(|_| "".to_string())
+        .to_ascii_lowercase();
+    if pretty == "1" || pretty == "true" {
```

Comment:
nit wrap this in a helper for reading from the env, a bit too imperative for my tastes

Reply:
I consider this as a permit to steal the lovely helper functions you wrote :D